GA Custom Domain Alerts

mkangia opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Make custom domain banners available to Advanced plan or higher ~along with it remaining accessible to projects added through feature flag so far.~
Deprecate the feature flag so it can be removed in a follow up PR post deploy of this.
PR to be merged only after a go ahead from necessary division(s) and once comms are ready/done, Hence added label "do not merge".

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SC-3198
Since we have moved features to GA before, using a previous [successful PR](https://github.com/dimagi/commcare-hq/pull/33078/files) as an example,
- Added migration that moves the feature from feature flag to a privilege for advanced and higher subscriptions
- Updated access checks

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
- Tested access changes locally. Feature available to domains ~that had the feature flag enabled and~ if privilege available (verified by test and by using "Test or Demo project" in UI).
- Migration was tested locally and on staging with QA

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
`corehq.apps.domain.tests.test_views:TestManageDomainAlertsView` & `corehq.apps.domain.tests.test_views:TestUpdateDomainAlertStatusView` &
`corehq.apps.users.tests.test_views:TestMobileWorkerListView` & 
`corehq.apps.users.tests.test_views:TestUpdateRoleFromView`

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi.atlassian.net/browse/QA-6004


### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [X] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

Reverting this PR will not cause any rollback on the records created/updated. Those extra records in the DB, will NOT conflict with the code deployed after the revert.
In case a permanent rollback of this PR is needed, then a new migration would be added to clean the redundant records so it can be done on all environments properly and not done manually.

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

mkangia committed:
converts ff to a privilege

mkangia committed:
updates subscriptions with new privilege

mkangia committed:
updates checks for feature access

mkangia committed:
updates permission management

mkangia committed:
refactor: DRY

mkangia committed:
reorders decorators

mkangia committed:
fixes failing tests

- fix copy by using deepcopy
- remove permission accessible by privilege

mkangia committed:
avoids running same test twice

mkangia committed:
Merge branch 'master' into mk/3198-ga-custom-alerts

zandre-eng commented:
> The migrations in this code can be safely applied first independently of the code.
This PR can be reverted after deploy with no further considerations.

These two ticked statements in the PR description aren't entirely correct. The migration does depend on the code as the code for the newly created privilege needs to exist for the migration to execute correctly. Additionally, if this PR were to be reverted, then the newly created role and grant records will need to be deleted from the DB to remove this new privilege.

zandre-eng commented:
One of the failing tests is as a result of [this line](https://github.com/dimagi/commcare-hq/blob/c68e58330cd2923969142bb214f67c844002721c/corehq/apps/domain/tests/test_views.py#L312) that needs to be updated to `@privilege_enabled(...)`

zandre-eng reviewed (approved):
Overall looks good from a code persective. Definitely a good idea to test this out on staging before merging, especially if this needs to go through UAT.

mkangia commented:
Hey @zandre-eng 

Thanks for the review

> The migration does depend on the code as the code for the newly created privilege needs to exist for the migration to execute correctly

A migration is run from the new code only during deploy. This checkbox on the PR ensures that the environment can continue to run on the old code even after the migration is applied in the background during deploy which changes the DB schema. So, that should be fine here. I tested by running the old code after running the migration locally.

> Additionally, if this PR were to be reverted, then the newly created role and grant records will need to be deleted from the DB to remove this new privilege.

Slight correction there. If the PR is reverted the newly created role will not be deleted. 
Firstly, we don't rollback migrations during deploy. So, if the PR is reverted and a new deploy is done, the migrations are not reverted.
Even if the migration was reverted manually, the newly created role will still not get deleted because we have `reverse_code=migrations.RunPython.noop`. 
In the PR, this checkbox tries to communicate that if there is an urgent revert of the PR, it can be done without any considerations. So, in this case if there is a revert of the PR, the permission records will stay in the DB but will not cause any failures. Usually, in such cases, the PR is added back by resolving the issue it caused. And in case, the PR needs to be permanently reverted then a new migration would need to be added to clean up records properly so it can be done on all environments. You should be hearing more about this in the next tech team meeting in cross divisional updated, this was discussed recently and we are trying to make this more clear across the team.

Let me know if that makes sense.

Was the migration applied differently in [this PR](https://github.com/dimagi/commcare-hq/pull/33078/files) by you or it went in the usual deploy?

mkangia commented:
> One of the failing tests is as a result of [this line](https://github.com/dimagi/commcare-hq/blob/c68e58330cd2923969142bb214f67c844002721c/corehq/apps/domain/tests/test_views.py#L312) that needs to be updated to @privilege_enabled(...)

Thanks for pointing that. I resolved a conflict last night that added some test that used the feature flag enable decorator from before. I will make that change.

mkangia committed:
updates decorator in tests

mkangia commented:
> Definitely a good idea to test this out on staging before merging, especially if this needs to go through UAT.

Good point about testing this on staging. The feature would go through UAT but this PR would not so it makes sense to either test this on staging or to go through QA.

I will submit a QA request

zandre-eng commented:
> Hey @zandre-eng
Thanks for the review
...

@mkangia Thanks for clarifying these points, that definitely makes sense. It would seem I had a slight misunderstanding with regards to what the checkboxes actually represented,

The migration in the linked PR went through the standard deploy process.

mkangia commented:
Thanks @zandre-eng 

> It would seem I had a slight misunderstanding with regards to what the checkboxes actually represented,

I totally understand. There is quite some context there and I am sure the update in the tech team meeting would be helpful to gain clarity.

mkangia committed:
updates to use the new valid check

mkangia committed:
fixes decorator call

mkangia committed:
merge master

mkangia committed:
deprecates toggle instead

mkangia commented:
Hey @zandre-eng 
There was a reconsideration about the final state of the feature flag and the conclusion was to remove it instead since we don't need to support it.

I have updated the description and added a change to deprecate the toggle in this PR. I will remove it in a follow up PR to keep the PR reversible.
Requesting for a review of the new changes.

zandre-eng commented on corehq/toggles/__init__.py:
```diff
@@ -2698,7 +2698,7 @@ def domain_has_privilege_from_toggle(privilege_slug, domain):
     slug='custom_domain_banners',
     label='Allow projects to add banners for their users on HQ',
-    tag=TAG_CUSTOM,
+    tag=TAG_DEPRECATED,
```
This was changed from a `FrozenPrivilegeToggle` back to a `StaticToggle`. Would the migrations for giving this privilege to relevant domains still work? Also, do all the privilege checks still work if this is a static toggle?

mkangia replied:
> Would the migrations for giving this privilege to relevant domains still work?

The migration is meant to just add the new privilege so there should be no change there. It is independent of the toggle. So, in case of supporting relevant domains during GA, the new privilege isn't passed over to relevant domains but the access is handled in code. Shared useful code links about that later in the comment. Let me know if anything seems off for you.

> Also, do all the privilege checks still work if this is a static toggle?

Privilege checks only fallback to frozen toggles and not static toggles when needed. So after this change, the toggle is simply archived (deprecated) and is not referred to anymore.

Some more useful context:

The logic that determines if the privilege falls back to a feature flag, is handled in code. It's in the function that I replaced, `requires_privilege_with_fallback` and additionally in the function `request_has_privilege`. Both of these eventually checks for only frozen type toggles through `domain_has_privilege_from_toggle`

https://github.com/dimagi/commcare-hq/blob/f134cf08a19f770123c37edcf853c135c5e17d19/corehq/toggles/__init__.py#L2543-L2545

and we get that collection from

https://github.com/dimagi/commcare-hq/blob/f134cf08a19f770123c37edcf853c135c5e17d19/corehq/toggles/__init__.py#L2536-L2539

`requires_privilege_with_fallback` was replaced with `requires_privilege_raise404` so it does not check for toggle anymore. And `request_has_privilege` checks for frozen toggles only, so the deprecated feature flag isn't referred to anymore.

zandre-eng replied:
Ah, thanks for the thorough explanation! I mistakenly thought the privilege was dependent on a `FrozenPrivilegeToggle`  to work correctly. Given that we aren't referring to the frozen privilege at all anymore, all good from my side!

zandre-eng replied:
Not blocking, but I'm curious if there was a specific reason for removing the feature flag entirely instead of converting it into a frozen privilege toggle? This was the norm for GA'ing features flags in the past, so I'd just like to understand the different approach taken here.

mkangia replied:
Hey @zandre-eng 

Yup, that is different from previous GA processes.
It is because we didn't have any projects that got access to the feature through feature flag and didn't have the necessary subscription to have access to the feature once it would become GA, so there was no need to retain the feature flag.
More in the comment [here](https://dimagi.atlassian.net/browse/SC-3198?focusedCommentId=308543)

zandre-eng reviewed (commented):
Overall looks good. I just have one blocking question.

mkangia commented:
Thanks @zandre-eng for the quick follow up.

zandre-eng reviewed (approved):
Looks good.
